File: app/src/main/java/org/baiyu/fuckshare/exifhelper/pngExifHelper.java
package org.baiyu.fuckshare.exifhelper;

import android.util.Log;

import org.baiyu.fuckshare.Utils;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Set;

public class pngExifHelper implements ExifHelper {

    private static final Set<String> pngCriticalChunks = Set.of(
            "acTL", //   animation control
            "bKGD", //   background color
            "cHRM", //   Primary Chromaticities
            "gAMA", //   Gamma
            "gIFg", //   GIFGraphicControlExtension
            "gIFt", //   GIFPlainTextExtension
            "gIFx", //   GIFApplicationExtension
            "fcTL", // * frame control
            "fdAT", // * frame data
            "hIST", //   PaletteHistogram
            "IDAT", // * image data
            "IEND", // * trailer
            "IHDR", // * header
            "pCAL", //   Pixel Calibration
            "pHYs", //   PhysicalPixel
            "PLTE", // * palette
            "sBIT", //   SignificantBits
            "sCAL", //   SubjectScale
            "sRGB", //   SRGBRendering
            "sTER", //   StereoImage
            "tRNS", //   Transparency
            "vpAg"  //   VirtualPage
    );

    @Override
    public void removeMetadata(InputStream inputStream, OutputStream outputStream) {
        try {

            BufferedInputStream bis;
            BufferedOutputStream bos;

            if (inputStream instanceof BufferedInputStream) {
                bis = (BufferedInputStream) inputStream;
            } else {
                bis = new BufferedInputStream(inputStream);
            }

            if (outputStream instanceof BufferedOutputStream) {
                bos = (BufferedOutputStream) outputStream;
            } else {
                bos = new BufferedOutputStream(outputStream);
            }

            byte[] byteArray = new byte[8];
            Utils.inputStreamRead(bis, byteArray);
            bos.write(byteArray);

            byte[] chunkLengthBytes = new byte[4];
            byte[] chunkNameBytes = new byte[4];
            long chunkDataCRCLength;

            while (bis.available() > 0) {
                Utils.inputStreamRead(bis, chunkLengthBytes);
                // 4 bytes of crc
                chunkDataCRCLength = Utils.bigEndianBytesToLong(chunkLengthBytes) + 4;

                Utils.inputStreamRead(bis, chunkNameBytes);
                String chunkName = new String(chunkNameBytes);

                if (pngCriticalChunks.contains(chunkName)) {
                    bos.write(chunkLengthBytes);
                    bos.write(chunkNameBytes);
                    chunkDataCRCLength -= Utils.copy(bis, bos, chunkDataCRCLength);
                } else {
                    // skip chunkData and chunkCrc
                    chunkDataCRCLength -= Utils.inputStreamSkip(bis, chunkDataCRCLength);
                    Log.d("fuckshare", "Discord chunk: " + chunkName + " size: " + chunkDataCRCLength + 4);
                }
                assert chunkDataCRCLength == 0;
                if (chunkName.equals("IEND")) {
                    break;
                }
            }
            bos.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/org/baiyu/fuckshare/exifhelper/PngExifHelper.kt
package org.baiyu.fuckshare.exifhelper

import org.baiyu.fuckshare.utils.ByteUtils
import org.baiyu.fuckshare.utils.ByteUtils.toUInt
import org.baiyu.fuckshare.utils.FileUtils
import timber.log.Timber
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteOrder

class PngExifHelper : ExifHelper {
    @Throws(ImageFormatException::class, IOException::class)
    override fun removeMetadata(inputStream: InputStream, outputStream: OutputStream) {
        val bis = inputStream.buffered()
        val bos = outputStream.buffered()

        FileUtils.copy(bis, bos, 8)

        val chunkLengthBytes = ByteArray(4)
        val chunkNameBytes = ByteArray(4)
        var chunkDataCRCLength: Long

        while (bis.available() > 0) {
            ByteUtils.readNBytes(bis, chunkLengthBytes)
            ByteUtils.readNBytes(bis, chunkNameBytes)
            // 4 bytes of crc
            chunkDataCRCLength = chunkLengthBytes.toUInt(ByteOrder.BIG_ENDIAN).toLong() + 4
            val chunkName = chunkNameBytes.toString(Charsets.US_ASCII)
            chunkDataCRCLength -= if (pngCriticalChunks.contains(chunkName)) {
                bos.write(chunkLengthBytes)
                bos.write(chunkNameBytes)
                Timber.d("Copy chunk: $chunkName size: ${chunkDataCRCLength + 4}")
                FileUtils.copy(bis, bos, chunkDataCRCLength)
            } else {
                // skip chunkData and chunkCrc
                Timber.d("Discord chunk: $chunkName size: ${chunkDataCRCLength + 4}")
                ByteUtils.skipNBytes(bis, chunkDataCRCLength)
            }
            if (chunkDataCRCLength != 0L) {
                throw ImageFormatException()
            }
            if (chunkName == "IEND") {
                break
            }
        }
        bos.flush()
    }

    companion object {
        private val pngCriticalChunks = setOf(
            "acTL",  //   animation control
            "bKGD",  //   background color
            "cHRM",  //   Primary Chromaticities
            "gAMA",  //   Gamma
            "gIFg",  //   GIFGraphicControlExtension
            "gIFt",  //   GIFPlainTextExtension
            "gIFx",  //   GIFApplicationExtension
            "fcTL",  // * frame control
            "fdAT",  // * frame data
            "hIST",  //   PaletteHistogram
            "IDAT",  // * image data
            "IEND",  // * trailer
            "IHDR",  // * header
            "pCAL",  //   Pixel Calibration
            "pHYs",  //   PhysicalPixel
            "PLTE",  // * palette
            "sBIT",  //   SignificantBits
            "sCAL",  //   SubjectScale
            "sRGB",  //   SRGBRendering
            "sTER",  //   StereoImage
            "tRNS",  //   Transparency
            "vpAg" //   VirtualPage
        )
    }
}